### ISSUE TITLE
Support Java 17

### ISSUE THREAD

**Issue by renatoathaydes** [jerkar/jeka#198]
Due to the way the classloader is currently changing the internal array of URLs in `URLClassLoader`, Jeka won't work in Java 17.

Error:

```
java.lang.reflect.InaccessibleObjectException: Unable to make protected void java.net.URLClassLoader.addURL(java.net.URL) accessible: module java.base does not "opens java.net" to unnamed module
```

As you probably know, doing that is extremely hacky :) there are much better ways to go about it... If you need help coming up with a proper way to handle dynamic classloading, let me know as I have some experience implementing that.

Not sure if anything else should break in Java 17, but given this issue was already causing a big warning in Java 9+, I suspect this is the only problem.

**Comment by renatoathaydes** [comment]
I had a look at the code base... unfortunately, the problem with loading classes is quite widespread.

Still, I think that to achieve the objectives of Jeka, you could make the class loading much more reliable (not to mention it won't break in Java 17 or future versions) by doing the following instead of using a mutable `JkClassLoader` that wraps `URLClassLoader`:

* use a Java parser library to collect all libraries you need to include in the build's classpath from the annotations in the build classes.
* Alternatively: install an annotation processor to do that as you compile all build classes.
* Remember the name of the parsed/processed classes (as we'll call them later from Jeka).
* Compile the build's Java sources and put them all in a jar file (or many if you need to isolate classpaths by class). 
* build an immutable `URLClassLoader` containing the URL of all jars (dependencies plus the compiled jar).
* Load the build classes by name, as necessary, with a simple `classLoader.load(className)`.

As I am not very familiar with the whole code base, I am probably missing something, but from the functionality I could gather that Jeka offers, this procedure should probably suffice, no?

**Comment by djeang** [comment]
Indeed, I was aware of this issue for jdk11.
I fixed it for 11 & 17 in branch 'rework engine'.
Note that this branch holds a completely rewritten engine. It is not mature nor compatible with current release.

The priority is now to release this new engine. Some work is still necessary as documentation and stabilization but it won't take long I hope.
If you're interested to have look on Jeka or even contribute to, it's the place to go.

**Comment by renatoathaydes** [comment]
FYI it works fine for me on JDK11, both from Azul and GraalVM.

**Comment by djeang** [comment]
resolved in 0.9.20.RC4
